---
title: ""
author: Hans Schrieke and Julie Reveillaud
date: "`r Sys.Date()`"
output: 
  rmdformats::readthedown:
    lightbox: true
    gallery: false
    fig_caption: true
    highlight: tango
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{css, echo=FALSE}
.main-container {
    max-width: 700px !important;
}

pre {
  max-height: 800px !important;
  overflow-y: auto !important;
  overflow-x: scroll !important;
}
pre code {
  white-space: pre
}
```

# Get split coverage and SNVs from selected split of Wolbachia O11 MAGs
```{zsh engine.opts='-i'}
# activate conda environment
conda activate anvio-7.1

# move to work directory
cd /Volumes/Elements/Hiseq/metapangenomics/output

# create specific folder for outputs
mkdir snv_figures


for s in sample_name M11 O11 O03 O07 O12
do
echo "$s" >> snv_figures/sample_of_interest.txt
done


for split in Culex_O11_000000098617_split_00001 Culex_O11_000000118914_split_00001 Culex_O11_000000120111_split_00001
do

# get coverage from splits of interest of O11 for SNV visualization
anvi-get-split-coverages -p 06_MERGED_references_mode/O11/PROFILE.db \
                         -c 03_CONTIGS_references_mode/O11-contigs.db \
                         --split-name ${split} \
                         -o snv_figures/${split}_coverage.txt

# generate SNV table from splits of interests of O11
echo ${split} > snv_figures/split_of_interest.txt

anvi-gen-variability-profile -p 06_MERGED_references_mode/O11/PROFILE.db \
                             -c 03_CONTIGS_references_mode/O11-contigs.db \
                             --splits-of-interest snv_figures/split_of_interest.txt \
                             --include-split-names \
                             --include-contig-names \
                             -o snv_figures/${split}_SNVs.txt
done

```

# Filter SNVs

```{r, warning=FALSE, message=FALSE}
# packages
require(tidyverse)

# import SNV table
setwd("/Volumes/Elements/Hiseq/metapangenomics/output/snv_figures")

splits <- c("Culex_O11_000000098617_split_00001_SNVs", "Culex_O11_000000118914_split_00001_SNVs", "Culex_O11_000000120111_split_00001_SNVs")

for(i in splits){
  snv <- read.table(paste0(i, ".txt"), header=TRUE)
  
  # filter the SNV table by removing the outliers
  snv_wt_outlier <- snv[snv$cov_outlier_in_contig==0 & snv$cov_outlier_in_split==0, ]
  
  # filter the SNV table by entropy and departure from consensus
  snv_entropy <- snv_wt_outlier[snv_wt_outlier$entropy>=0.2,]
  snv_dep <- snv_entropy[snv_entropy$departure_from_consensus>=0.2,]
  
  # export snv tables
  write.table(snv_wt_outlier, paste0(i, "_wt_outliers.txt"), sep="\t", quote=FALSE, row.names=FALSE)
  write.table(snv_dep, paste0(i, "_filtered.txt"), sep="\t", quote=FALSE, row.names=FALSE)
}
```

# Generate figures with filtered SNV tables
```{zsh engine.opts='-i'}
# activate conda environment
conda activate anvio-7.1

# move to work directory
cd /Volumes/Elements/Hiseq/metapangenomics/output/snv_figures

for split in Culex_O11_000000098617_split_00001 Culex_O11_000000118914_split_00001 Culex_O11_000000120111_split_00001
do

# generate pdf of split of interest (without SNVs)                        
anvi-script-visualize-split-coverages -i ${split}_coverage.txt \
                                      -o ${split}_inspect.pdf

# generate pdf of splits of interest with SNVs
anvi-script-visualize-split-coverages -i ${split}_coverage.txt \
                                      -o ${split}_inspect_SNV.pdf \
                                      --snv-data ${split}_SNVs.txt \
                                      -s sample_of_interest.txt \
                                      --free-y-scale TRUE

# generate pdf of splits of interest with non-outliers SNVs
anvi-script-visualize-split-coverages -i ${split}_coverage.txt \
                                      -o ${split}_inspect_SNV_wt_outliers.pdf \
                                      --snv-data ${split}_SNVs_wt_outliers.txt \
                                      -s sample_of_interest.txt \
                                      --free-y-scale TRUE
                                      
# generate pdf of splits of interest with filtered SNVs
anvi-script-visualize-split-coverages -i ${split}_coverage.txt \
                                      -o ${split}_inspect_SNV_filtered.pdf \
                                      --snv-data ${split}_SNVs_filtered.txt \
                                      -s sample_of_interest.txt \
                                      --free-y-scale TRUE
done
```

# Generate gene figures for each split of interest

```{r, warning=FALSE, message=FALSE}
library(gggenes)
library(tidyverse)

setwd("/Volumes/Elements/Hiseq/metapangenomics/output/snv_figures")

test <- example_genes

gene_calls <- read.table("/Volumes/Elements/Hiseq/metapangenomics/output/03_CONTIGS_references_mode/O11-SUMMARY/bin_by_bin/O11/O11-gene_calls.txt", header=TRUE, fill=TRUE, sep="\t", quote = "")

gene_calls_selected <- gene_calls[gene_calls$contig=="Culex_O11_000000098617" | gene_calls$contig=="Culex_O11_000000118914" | gene_calls$contig=="Culex_O11_000000120111",]
gene_calls_selected <- gene_calls_selected %>% select(c(gene_callers_id, start, stop, direction, contig, COG20_FUNCTION))
gene_calls_selected$color <- "green"
gene_calls_selected[gene_calls_selected$COG20_FUNCTION=="", "color"] <- "grey"
gene_calls_selected <- gene_calls_selected %>% select(c(gene_callers_id, start, stop, direction, contig, color))


colnames(gene_calls_selected) <- c("gene", "start", "end", "strand", "molecule", "color")
levels(gene_calls_selected$strand) <- c("forward", "reverse")
gene_calls_selected$orientation <- 1
gene_calls_selected[gene_calls_selected$strand=="reverse", "orientation"] <- -1

gene_calls_selected$gene <- gene_calls_selected$gene %>% as.character()

p <- ggplot(gene_calls_selected, aes(xmin = start, xmax = end, y = molecule, fill = gene, forward=orientation, label=gene)) +
  geom_gene_arrow()+
  facet_wrap(~ molecule, scales = "free", ncol = 1)+
  geom_gene_label(align = "centre") +
  scale_fill_manual(values=gene_calls_selected$color) +
  theme_genes()

ggsave("genes_inspect.pdf",p, width=4, height=3, units="in", scale=3)
```
