---
title: "SNV Analysis in Wolbachia from midguts and ovaries mosquitoes "
author: Hans Schrieke and Julie Reveillaud
date: "`r Sys.Date()`"
output: 
  rmdformats::readthedown:
    lightbox: true
    gallery: false
    fig_caption: true
    highlight: tango
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{css, echo=FALSE}
.main-container {
    max-width: 700px !important;
}

pre {
  max-height: 800px !important;
  overflow-y: auto !important;
  overflow-x: scroll !important;
}
pre code {
  white-space: pre
}
```

# SNV tables

## samples of interest file

```{bash}
mkdir ../../output/SNV ../../output/SNV/snv-tables ../../output/SNV/snv-tables/intra-MAGs ../../output/SNV/snv-tables/intra-MAGs/state-files ../../output/SNV/snv-tables/SAAVs ../../output/SNV/snv-tables/SCVs ../../output/SNV/snv-tables/indels

rm ../../output/SNV/snv-tables/sample-of-interest.txt

for s in O03 M03 O07 M07 O11 M11 O12 M12
do
echo "$s" >> ../../output/SNV/snv-tables/sample-of-interest.txt
done

for s in O03 O07 O11 M11 O12
do
mkdir ../../output/SNV/snv-tables/${s}-as-reference
cp ../../output/SNV/snv-tables/sample-of-interest.txt ../../output/SNV/snv-tables/${s}-as-reference/sample-of-interest.txt
#cat ../../output/SNV/snv-tables/${s}-as-reference/sample-of-interest.txt
done

rm ../../output/SNV/snv-tables/sample-of-interest.txt 
```


## SNV (INTER-MAGs - ALL METAGENOMES ON EACH METAGENOME AS REFERENCE)
```{zsh engine.opts='-i'}
conda activate anvio-7.1

for s in O03 O07 O11 M11 O12
do

    ## SNVs
    anvi-gen-variability-profile -p ../../output/06_MERGED_references_mode/${s}/PROFILE.db  \
                                 -c ../../output/03_CONTIGS_references_mode/${s}-contigs.db \
                                 -C Wolbachia \
                                 -b ${s} \
				                         --samples-of-interest ../../output/SNV/snv-tables/${s}-as-reference/sample-of-interest.txt \
			                           --include-split-names \
                                 --include-contig-names \
                                 --compute-gene-coverage-stats \
                                 -o ../../output/SNV/snv-tables/${s}-as-reference/variability-${s}-all.txt

    ## SAAVs
    anvi-gen-variability-profile -p ../../output/06_MERGED_references_mode/"$s"/PROFILE.db \
                                 -c ../../output/03_CONTIGS_references_mode/"$s"-contigs.db \
                                 -C Wolbachia \
                                 -b "$s" \
                                 --engine AA \
                                 --include-split-names \
                                 --samples-of-interest ../../output/SNV/snv-tables/${s}-as-reference/sample-of-interest.txt \
                                 -o ../../output/SNV/snv-tables/SAAVs/SAAV_variability_"$s"_inter-MAGs.txt
      
    ## SCVs  
    anvi-gen-variability-profile -p ../../output/06_MERGED_references_mode/"$s"/PROFILE.db \
                                 -c ../../output/03_CONTIGS_references_mode/"$s"-contigs.db \
                                 -C Wolbachia \
                                 -b "$s" \
                                 --engine CDN \
                                 --include-split-names \
                                 --samples-of-interest ../../output/SNV/snv-tables/${s}-as-reference/sample-of-interest.txt \
                                 -o ../../output/SNV/snv-tables/SCVs/SCV_variability_"$s"_inter-MAGs.txt
    
    # ## indels
    # anvi-export-table ../../output/06_MERGED_references_mode/"$s"/PROFILE.db --table ../../output/SNV/snv-tables/indels/indels-"$s".txt

done

conda deactivate
```

### Filter SNV table (inter-MAGs)

```{r, warning=FALSE}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")

for(i in samples){
  snv <- read.table(paste0("../../output/SNV/snv-tables/", i, "-as-reference/variability-", i, "-all.txt"), header=TRUE)
  
  # remove outliers snv
  snv_wt_outlier <- snv[snv$cov_outlier_in_contig==0 & snv$cov_outlier_in_split==0, ]
  
  # filter entropy
  snv_entropy <- snv_wt_outlier[snv_wt_outlier$entropy>=0.2 & snv_wt_outlier$departure_from_reference>=0.2,]
  
  # save tables
  write.table(snv_entropy, paste0("../../output/SNV/snv-tables/", i, "-as-reference/variability-", i, "-filtered.txt"), sep="\t", quote=FALSE, row.names=FALSE)
}
```

### Filter SCV table (inter-MAGs)

```{r, warning=FALSE}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")

getwd()

for(i in samples){
  scv <- read.table(paste0("../../output/SNV/snv-tables/SCVs/SCV_variability_", i, "_inter-MAGs.txt"),
                     header=TRUE, fill=TRUE, sep="\t", quote = "")
  
  # filter entropy
  scv_entropy <- scv[scv$entropy>=0.2 & scv$departure_from_reference>=0.2,]
  
  # save tables
  write.table(scv_entropy, paste0("../../output/SNV/snv-tables/SCVs/SCV_variability_", i, "_filtered.txt"), sep="\t", quote=FALSE, row.names=FALSE)
}
```

### Filter SAAV table (inter-MAGs)

```{r, warning=FALSE}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")

for(i in samples){
  saav <- read.table(paste0("../../output/SNV/snv-tables/SAAVs/SAAV_variability_", i, "_inter-MAGs.txt"),
                     header=TRUE, fill=TRUE, sep="\t", quote = "")
  
  # filter entropy
  saav_entropy <- saav[saav$entropy>=0.2 & saav$departure_from_reference>=0.2,]
  
  # save tables
  write.table(saav_entropy, paste0("../../output/SNV/snv-tables/SAAVs/SAAV_variability_", i, "_filtered.txt"), sep="\t", quote=FALSE, row.names=FALSE)
}
```

## SNV (INTRA-MAGs - EACH METAGENOME ON EACH CORRESPONDING MAG)

```{zsh engine.opts='-i'}
conda activate anvio-7.1

# generate variability profiles for each Wolbachia genome 
for s in O03 O07 O11 O12 M11
do
    # ## SNVs
    echo "$s" > ../../output/SNV/snv-tables/intra-MAGs/sample-of-interest-MAGs.txt
    anvi-gen-variability-profile -p ../../output/06_MERGED_references_mode/"$s"/PROFILE.db \
                                 -c ../../output/03_CONTIGS_references_mode/"$s"-contigs.db \
                                 -C Wolbachia \
                                 -b "$s" \
                                 --include-split-names \
                                 --samples-of-interest ../../output/SNV/snv-tables/intra-MAGs/sample-of-interest-MAGs.txt \
                                 -o ../../output/SNV/snv-tables/intra-MAGs/variability_"$s".txt
    
    ## SAAVs
    anvi-gen-variability-profile -p ../../output/06_MERGED_references_mode/"$s"/PROFILE.db \
                                 -c ../../output/03_CONTIGS_references_mode/"$s"-contigs.db \
                                 -C Wolbachia \
                                 -b "$s" \
                                 --engine AA \
                                 --include-split-names \
                                 --samples-of-interest ../../output/SNV/snv-tables/intra-MAGs/sample-of-interest-MAGs.txt \
                                 -o ../../output/SNV/snv-tables/SAAVs/SAAV_variability_"$s"_intra-MAGs.txt
      
    ## SCVs  
    anvi-gen-variability-profile -p ../../output/06_MERGED_references_mode/"$s"/PROFILE.db \
                                 -c ../../output/03_CONTIGS_references_mode/"$s"-contigs.db \
                                 -C Wolbachia \
                                 -b "$s" \
                                 --engine CDN \
                                 --include-split-names \
                                 --samples-of-interest ../../output/SNV/snv-tables/intra-MAGs/sample-of-interest-MAGs.txt \
                                 --include-site-pnps \
                                 -o ../../output/SNV/snv-tables/SCVs/SCV_variability_"$s"_intra-MAGs.txt
done

# combine all variability information into a single file:
head -n 1 ../../output/SNV/snv-tables/intra-MAGs/variability_O03.txt  > ../../output/SNV/snv-tables/intra-MAGs/variability-MAGs-all.txt
cat ../../output/SNV/snv-tables/intra-MAGs/variability_* | grep -v entry_id >> ../../output/SNV/snv-tables/intra-MAGs/variability-MAGs-all.txt

# generate a text file with genome lenghts of Wolbachia and samples
# from which they were generated:
echo -e "metagenome\tgenome_length" > ../../output/SNV/snv-tables/intra-MAGs/genome-lengths.txt
for s in M11 O03 O07 O11 O12
do
    len=`sqlite3 ../../output/03_CONTIGS_references_mode/"$s"-contigs.db 'select value from self where key=="total_length";'`
    echo -e "Culex_$s\t$len" >> ../../output/SNV/snv-tables/intra-MAGs/genome-lengths.txt
done

conda deactivate

```



### Filter SNV table (intra-MAGs)

```{r}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")

for(i in samples){
  snv <- read.table(paste0("/Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/intra-MAGs/variability_", i, ".txt"), header=TRUE)
  
  # # remove outliers snv
  snv_wt_outlier <- snv[snv$cov_outlier_in_contig==0 & snv$cov_outlier_in_split==0, ]
  
  # filter entropy
  snv_entropy <- snv_wt_outlier[snv_wt_outlier$entropy>=0.2,]
  
  # filter departure from consensus
  snv_dep <- snv_entropy[snv_entropy$departure_from_consensus>=0.2,]
  
  # # prepare layers to add in profile.db of interactive
  # snv_layers <- snv_entropy %>% select(c(entry_id, unique_pos_identifier, pos, pos_in_contig, split_name, in_coding_gene_call, coverage, entropy))
  snv_layers <- snv_entropy %>% select(c(entry_id, split_name, in_coding_gene_call, coverage, entropy))
  snv_layers$entry_id <- paste0("p_", sprintf("%08d", as.numeric(snv_layers$entry_id)))
  
  
  # write filtered snv table
  ## table for anvi-script-snvs-to-interactive
  write.table(snv_dep, paste0("/Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/intra-MAGs/variability-", i, "-filtered-2.txt"), sep="\t", quote=FALSE, row.names=FALSE)
  
  ## layers to add in snvs-interactive
  write.table(snv_layers, paste0("/Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/intra-MAGs/variability-", i, "-layers.txt"), sep="\t", quote=FALSE, row.names=FALSE)

}
  
```



### Filter SCV table (intra-MAGs)

```{r}
require(tidyverse)

samples <- c("O11","O03", "O07", "O12", "M11")

for(i in samples){
  scv <- read.table(paste0("/Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/SCVs/SCV_variability_", i, "_intra-MAGs.txt"), header=TRUE, fill=TRUE)
  
  # filter entropy
  scv_entropy <- scv[scv$entropy>=0.2,]
  
  # filter departure from consensus
  scv_dep <- scv_entropy[scv_entropy$departure_from_consensus>=0.2,]
  
  write.table(scv_dep, paste0("/Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/SCVs/SCV_variability_", i, "_intra-MAGs_filtered.txt"), sep="\t", quote=FALSE, row.names=FALSE)
}
  
```


### Filter SAAV table (intra-MAGs)

```{r}
require(tidyverse)

samples <- c("O11","O03", "O07", "O12", "M11")

for(i in samples){
  saav <- read.table(paste0("/Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/SAAVs/SAAV_variability_", i, "_intra-MAGs.txt"), header=TRUE, fill=TRUE)
  
  # filter entropy
  saav_entropy <- saav[saav$entropy>=0.2,]
  
  # filter departure from consensus
  saav_dep <- saav_entropy[saav_entropy$departure_from_consensus>=0.2,]
  
  write.table(saav_dep, paste0("/Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/SAAVs/SAAV_variability_", i, "_intra-MAGs_filtered.txt"), sep="\t", quote=FALSE, row.names=FALSE)
}
  
```

# Indel tables
```{zsh engine.opts='-i'}
cd ../../output/SNV/snv-tables/indels

conda activate anvio-7.1

for s in O03 O07 O11 M11 O12
do

## indels
anvi-export-table ../../../../output/06_MERGED_references_mode/"$s"/PROFILE.db --table indels -o "$s"_indels.txt

done

conda deactivate
```

## Filter indels

```{r, warning=FALSE}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")

for(i in samples){
  indels <- read.table(paste0("../../output/SNV/snv-tables/indels/", i, "_indels.txt"), header=TRUE, fill=TRUE, sep="\t", quote = "")
  
  # remove outliers snv
  indels_wt_outlier <- indels[indels$cov_outlier_in_contig==0 & indels$cov_outlier_in_split==0, ]
  
  # save tables
  write.table(indels_wt_outlier, paste0("../../output/SNV/snv-tables/indels/", i, "_indels_filtered.txt"), sep="\t", quote=FALSE, row.names=FALSE)
}
```

# SNV to interactive

```{zsh engine.opts='-i'}
conda activate anvio-7.1

cd /Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/intra-MAGs

# for each sample
for i in O03 O07 O11 O12 M11
do

  ## convert snv table to interactive

  anvi-script-snvs-to-interactive variability-$i-filtered-2.txt \
                                  -o ${i}_wolbachia_filtered \
                                  --min-departure-from-consensus 0.2

  ## add layers to interactive
  anvi-import-misc-data variability-$i-layers.txt \
                        -p ${i}_wolbachia_filtered/profile.db \
                        --target-data-table items \
                        --just-do-it

done
```

## Export the state files

After draw a nice plot in interactive for O11, we can export the state file and adpapt him for each other sample

```{zsh eval=FALSE, engine.opts='-i', include=FALSE}
conda activate anvio-7.1

cd /Volumes/Elements/Hiseq/metapangenomics/output/SNV/snv-tables/intra-MAGs

for i in O03 O07 O11 O12 M11
do

anvi-import-state -s /Volumes/Elements/Hiseq/metapangenomics/metadata/state-files/state-draft3-${i}.json \
                  -n draft3 \
                  -p ${i}_wolbachia_filtered/profile.db

anvi-import-state -s /Volumes/Elements/Hiseq/metapangenomics/metadata/state-files/state-draft4-${i}.json \
                  -n draft4 \
                  -p ${i}_wolbachia_filtered/profile.db

anvi-import-state -s /Volumes/Elements/Hiseq/metapangenomics/metadata/state-files/state-draft4-gene-${i}.json \
                  -n draft4_gene \
                  -p ${i}_wolbachia_filtered/profile.db

anvi-import-state -s /Volumes/Elements/Hiseq/metapangenomics/metadata/state-files/state-draft4-snv-shared-${i}.json \
                  -n draft4_snv_shared \
                  -p ${i}_wolbachia_filtered/profile.db

anvi-import-state -s /Volumes/Elements/Hiseq/metapangenomics/metadata/state-files/state-draft4-gene-snv-shared-${i}.json \
                  -n draft4_gene_snv_shared \
                  -p ${i}_wolbachia_filtered/profile.db

anvi-import-state -s /Volumes/Elements/Hiseq/metapangenomics/metadata/state-files/state-draft5-gene-${i}.json \
                  -n draft5_gene \
                  -p ${i}_wolbachia_filtered/profile.db

anvi-import-state -s /Volumes/Elements/Hiseq/metapangenomics/metadata/state-files/state-draft5-gene-snv-shared${i}.json \
                  -n draft5_gene_snv_shared \
                  -p ${i}_wolbachia_filtered/profile.db
done
```



## Display snv to interactive

```{zsh eval=FALSE, engine.opts='-i', include=FALSE}
conda activate anvio-7.1
anvi-interactive --profile O11_wolbachia_filtered/profile.db \
                 --tree O11_wolbachia_filtered/tree.txt \
                 --view-data O11_wolbachia_filtered/view.txt \
                 --title "SNV Profile for the O11 Wolbachia bin" \
                 --manual

anvi-interactive --profile M11_wolbachia_filtered/profile.db \
                 --tree M11_wolbachia_filtered/tree.txt \
                 --view-data M11_wolbachia_filtered/view.txt \
                 --title "SNV Profile for the M11 Wolbachia bin" \
                 --manual

anvi-interactive --profile O03_wolbachia_filtered/profile.db \
                 --tree O03_wolbachia_filtered/tree.txt \
                 --view-data O03_wolbachia_filtered/view.txt \
                 --title "SNV Profile for the O03 Wolbachia bin" \
                 --manual

anvi-interactive --profile O07_wolbachia_filtered/profile.db \
                 --tree O07_wolbachia_filtered/tree.txt \
                 --view-data O07_wolbachia_filtered/view.txt \
                 --title "SNV Profile for the O07 Wolbachia bin" \
                 --manual

anvi-interactive --profile O12_wolbachia_filtered/profile.db \
                 --tree O12_wolbachia_filtered/tree.txt \
                 --view-data O12_wolbachia_filtered/view.txt \
                 --title "SNV Profile for the O12 Wolbachia bin" \
                 --manual
```





