---
title: "6-generate-variability-tables"
author: "Hans Schrieke, Blandine Trouche and Julie Reveillaud"
date: "`r Sys.Date()`"
output: 
  github_document:
    toc: true
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/btrouche/Dropbox/RosaLind/SNV_Meren")
```

## Generate raw variability tables

```{bash tables, eval = FALSE}
# Create output directories
mkdir -p 07_RAW_VAR_TABLES
mkdir -p 08_FILTERED_VAR_TABLES

# Generate SNV tables 
conda activate anvio-7.1

for s in O03 O07 O11 M11 O12
do

    ## SNVs
    anvi-gen-variability-profile -p 06_MERGED_references_mode/${s}_filtered/PROFILE.db  \
                                 -c 03_CONTIGS_references_mode/${s}-contigs.db \
                                 -C Wolbachia \
                                 -b ${s} \
			                           --include-split-names \
                                 --include-contig-names \
                                 --compute-gene-coverage-stats \
                                 -o 07_RAW_VAR_TABLES/SNV_${s}.txt

    ## SAAVs
    anvi-gen-variability-profile -p 06_MERGED_references_mode/${s}_filtered/PROFILE.db \
                                 -c 03_CONTIGS_references_mode/${s}-contigs.db \
                                 -C Wolbachia \
                                 -b ${s} \
                                 --engine AA \
                                 --include-split-names \
                                 -o 07_RAW_VAR_TABLES/SAAV_${s}.txt
      
    ## SCVs  
    anvi-gen-variability-profile -p 06_MERGED_references_mode/${s}_filtered/PROFILE.db \
                                 -c 03_CONTIGS_references_mode/${s}-contigs.db \
                                 -C Wolbachia \
                                 -b ${s} \
                                 --engine CDN \
                                 --include-split-names \
                                 --include-site-pnps \
                                 -o 07_RAW_VAR_TABLES/SCV_${s}.txt

done

conda deactivate
```

## Add SCG info to the variability tables

### SNV tables
```{r, warning=FALSE}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")
#setwd("/Users/btrouche/Dropbox/RosaLind/SNV_Meren")

gene_clusters <- read.table("pangenomics/Wolbachia-all-PAN-SUMMARY/Wolbachia_gene_clusters_summary.txt", 
                            header=TRUE, fill=TRUE, sep="\t", quote = "")
mlst <- read.table("pangenomics/coverages-for-gene-clusters.txt", 
                   header=TRUE, fill=TRUE, sep="\t", quote = "")
mlst <- mlst %>% select(c(1:9))

for(i in samples){
  snv <- read.table(paste0("07_RAW_VAR_TABLES/SNV_", i,".txt"), header=TRUE)
  
  # raw intra-sample table 
  snv_intra <- snv[snv$sample== paste0(i, "_filter20"), ]

  # remove outliers snv
  snv_wt_outlier <- snv[snv$cov_outlier_in_contig==0 & snv$cov_outlier_in_split==0, ]
  
  # filter entropy
  snv_entropy <- snv_wt_outlier[snv_wt_outlier$entropy>=0.2 & snv_wt_outlier$departure_from_consensus>=0.2,]
  
  # prepare layers to add in profile.db of interactive
  snv_layers <- snv_entropy
  snv_layers$entry_id <- paste0("p_", sprintf("%08d", as.numeric(snv_layers$unique_pos)))
  
  # add SCG info
  gene_clusters_sample <- gene_clusters[gene_clusters$genome_name==i,]
  colnames(gene_clusters_sample)[5] <- "corresponding_gene_call"
  snv_layers <- snv_layers %>% merge(gene_clusters_sample, by="corresponding_gene_call", all.x = TRUE)
  snv_layers$corresponding_gene_call <- snv_layers$corresponding_gene_call %>% as.character()
  snv_layers <- snv_layers %>% select(c(entry_id, everything()))
  
  # add MLST + wsp + cid info
  snv_layers <- snv_layers %>% merge(mlst, by="gene_cluster_id", all.x = TRUE)
  
  # intra-sample
  snv_entropy_sample <- snv_entropy[snv_entropy$sample==paste0(i, "_filter20"), ]
  snv_layers_sample <- snv_layers[snv_layers$sample==paste0(i, "_filter20"), ]
  
  # change in coding gene call layer
  snv_layers_sample[snv_layers_sample$corresponding_gene_call != -1 & snv_layers_sample$in_coding_gene_call==0, "in_coding_gene_call"] <- 2
  snv_layers_sample$gene_coding <- snv_layers_sample$in_coding_gene_call %>% as.character() %>% as.factor()
  snv_layers_sample$gene_coding %>% levels()
  levels(snv_layers_sample$gene_coding) <- c("no coding gene", "coding gene", "partial gene")
  
  # change SCG layer
  snv_layers_sample$SCG_proper <- snv_layers_sample$SCG%>% as.character() %>% as.factor()
  levels(snv_layers_sample$SCG_proper) <- c("No", "Yes", "Out of gene")
  snv_layers_sample[is.na(snv_layers_sample$SCG_proper), "SCG_proper"] <- "Out of gene"

  #### ALSO ADD SCG INFO TO THE RAW TABLE, FOR SNP DETECTION
  snv_SCG <- snv
  snv_SCG$entry_id <- paste0("p_", sprintf("%08d", as.numeric(snv_SCG$unique_pos)))
  
  gene_clusters_sample <- gene_clusters[gene_clusters$genome_name==i,]
  colnames(gene_clusters_sample)[5] <- "corresponding_gene_call"
  snv_SCG <- snv_SCG %>% merge(gene_clusters_sample, by="corresponding_gene_call", all.x = TRUE)
  snv_SCG$corresponding_gene_call <- snv_SCG$corresponding_gene_call %>% as.character()
  snv_SCG <- snv_SCG %>% select(c(entry_id, everything()))
  
  # add MLST + wsp + cid info
  snv_SCG <- snv_SCG %>% merge(mlst, by="gene_cluster_id", all.x = TRUE)
  ####
  
  # save tables
  write.table(snv_layers_sample, paste0("08_FILTERED_VAR_TABLES/SNV/SNV_", i, "_intra_filtered_SCG.txt"), 
              sep="\t", quote=FALSE, row.names=FALSE)
  
   write.table(snv_SCG, paste0("07_RAW_VAR_TABLES/SNV_", i, "_SCG.txt"), 
              sep="\t", quote=FALSE, row.names=FALSE)
}
```

### SCV tables
```{r, warning=FALSE}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")

# add SCG info
for(i in samples){
  scv <- read.table(paste0("07_RAW_VAR_TABLES/SCV_", i,".txt"),
                    header=TRUE, fill=TRUE, sep="\t", quote = "")
  
  # prepare layers to add in profile.db of interactive
  scv_layers <- scv
  scv_layers$entry_id <- paste0("p_", sprintf("%08d", as.numeric(scv_layers$entry_id)))
  
  # add SCG info
  gene_clusters_sample <- gene_clusters[gene_clusters$genome_name==i,]
  colnames(gene_clusters_sample)[5] <- "corresponding_gene_call"
  scv_layers <- scv_layers %>% merge(gene_clusters_sample, by="corresponding_gene_call", all.x = TRUE)
  scv_layers$corresponding_gene_call <- scv_layers$corresponding_gene_call %>% as.character()
  scv_layers <- scv_layers %>% select(c(entry_id, everything()))
  
  # add MLST + wsp + cid info
  scv_layers <- scv_layers %>% merge(mlst, by="gene_cluster_id", all.x = TRUE)
  
  # save tables
  write.table(scv_layers, paste0("07_RAW_VAR_TABLES/SCV_", i, "_SCG.txt"), 
              sep="\t", quote=FALSE, row.names=FALSE)
}
```

### SAAV tables
```{r, warning=FALSE}
require(tidyverse)
samples <- c("O03", "O07", "O11", "O12", "M11")

# inter-sample level
for(i in samples){
  saav <- read.table(paste0("07_RAW_VAR_TABLES/SAAV_", i,".txt"),
                     header=TRUE, fill=TRUE, sep="\t", quote = "")
  
  # prepare layers to add in profile.db of interactive
  saav_layers <- saav
  saav_layers$entry_id <- paste0("p_", sprintf("%08d", as.numeric(saav_layers$entry_id)))
  
  # add SCG info
  gene_clusters_sample <- gene_clusters[gene_clusters$genome_name==i,]
  colnames(gene_clusters_sample)[5] <- "corresponding_gene_call"
  saav_layers <- saav_layers %>% merge(gene_clusters_sample, by="corresponding_gene_call", all.x = TRUE)
  saav_layers$corresponding_gene_call <- saav_layers$corresponding_gene_call %>% as.character()
  saav_layers <- saav_layers %>% select(c(entry_id, everything()))
  
  # add MLST + wsp + cid info
  saav_layers <- saav_layers %>% merge(mlst, by="gene_cluster_id", all.x = TRUE)
  
  # save tables
  write.table(saav_layers, paste0("07_RAW_VAR_TABLES/SAAV_", i, "_SCG.txt"), 
              sep="\t", quote=FALSE, row.names=FALSE)
}
```

